### PR TITLE
(MODULES-6844) Add v2-only helper

### PR DESCRIPTION
**Pull request by michaeltlombardi**
These commits add a V2-only helper by cloning and updating the adapter. It adds integration tests for the new helper and strips away pieces of the helper which are for V1 tasks.

It is _out of scope_ to wire this helper up to the provider or otherwise alter the type/provider for this helper.

### CODE REVIEW

**Comment by michaeltlombardi** [comment]
I can't think of much else to rip at this stage from the V2 helper. Some of the logic lives in the trigger class and the indexing stuff is in the provider, both of which are out of scope for this ticket.

**Comment by glennsarti** [comment]
Should there be unit tests for this or integration only?

**Comment by michaeltlombardi** [comment]
@glennsarti Possibly - I didn't see any unit tests for the adapter code, so I didn't add any new ones. I can noodle on adding some, but my original thought was to update the provider unit tests in the ticket to wire up the new helper to the provider.

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 1]
```diff
@@ -259,6 +259,7 @@ def self.append_v1trigger(definition, v1trigger)
       end
 
       # Values for all Trigger Types
+      iTrigger.RandomDelay         = "PT#{v1trigger['random_minutes_interval']}M" unless v1trigger['random_minutes_interval'].nil?   || v1trigger['random_minutes_interval'].zero?
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Could you fill in the docs on this commit as well.  There's nothing anywhere that mentions a user may set this.

Comment (michaeltlombardi) [reply lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Prior to this being hooked up with the provider, they definitely cannot. Where would that doc update land in this workset? Documenting it in this helper seems like the wrong place but we've pushed off updating the type/provider in this ticket for the wire-up ticket.

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    result = `#{full_cmd}`
+
+    result.strip
+  end
+
+  match do |actual|
+    from_ps = run_ps(ps_cmd)
+
+    # This matcher probably won't tolerate UTF8 characters
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: You can remove that comment... `==` should work fine for Unicode(*).


(*) There are some cases where Ruby can have problems - but its typically only when comparing normalized to non-normalized Unicode, which you wouldn't have to worry about here. See http://blog.honeybadger.io/ruby-s-unicode-support/

Comment (Iristyle) [reply spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: Also, if you're concerned about the output coming back from PS itself ... multiple responses would come back as the same sequence of bytes (if corrupted by capturing things over `stdout`), so should also match for that reason.

**Comment by Iristyle** [comment]
I think we might want to add the commit that exposes the task compatibility and then use that in tests?  Something similar to https://github.com/puppetlabs/puppetlabs-scheduled_task/pull/19/commits/9e10ee5fd06d438e2510c1a441bba1eb130fe6b8?

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb (hunk 2 above)
Comment (michaeltlombardi) [reply spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: Okay, I pulled this comment over when I stole the method itself from Glenn's work on the adapter integration tests.

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      'type'         => {
+        'weeks_interval' => 1,
+        'days_of_week'   => Win32::TaskScheduler::MONDAY,
+      },
+      'trigger_type'            => :TASK_TIME_TRIGGER_WEEKLY,
+    }),
+    defaults.merge({
+      'type'         => {
+        'months' => Win32::TaskScheduler::JANUARY,
+        # TODO: this value seems to be cooked up wrong
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: We still have this test TODO here which should probably be removed / sorted out...

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      'trigger_type'            => :TASK_TIME_TRIGGER_ONCE,
+    }),
+    defaults.merge({
+      'type'                    => { 'days_interval' => 1 },
+      'trigger_type'            => :TASK_TIME_TRIGGER_DAILY,
+    }),
+    defaults.merge({
+      'type'         => {
+        'weeks_interval' => 1,
+        'days_of_week'   => Win32::TaskScheduler::MONDAY,
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: I believe this constant now lives in the Trigger class... but might not yet until my next Trigger PR goes in

**Comment by Iristyle** [comment]
Do we wan to modify https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/spec/unit/puppet/provider/scheduled_task/win32_taskscheduler_spec.rb#L124 so that it also uses `TaskScheduler2Task` in addition to `TaskScheduler2V1Task`... I think that as we discussed, this can be deferred until the new helper is integrated with the provider.

File: lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task.rb [hunk 5]
```diff
@@ -36,7 +36,7 @@ def initialize(work_item = nil, trigger = nil)
   def enum
     @tasksched.enum_task_names(PuppetX::PuppetLabs::ScheduledTask::TaskScheduler2::ROOT_FOLDER,
       include_child_folders: false,
-      include_compatibility: [PuppetX::PuppetLabs::ScheduledTask::TaskScheduler2::TASK_COMPATIBILITY_AT, PuppetX::PuppetLabs::ScheduledTask::TaskScheduler2::TASK_COMPATIBILITY_V1]).map do |item|
+      include_compatibility: [PuppetX::PuppetLabs::ScheduledTask::TaskScheduler2::TASK_COMPATIBILITY_V2]).map do |item|
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task.rb]: This is probably the commit where we want some tests to at least vet the compatibility on the task returned.

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb (hunk 1 above)
Comment (Iristyle) [reply lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Yeah, that's a good point.  Of course, users won't really get their hands on the module until the code actually ships, at which point the code will be integrated.

I don't feel super strongly either way, but it's usually helpful to update docs when a code change lands that enables a feature?

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb (hunk 4 above)
Comment (michaeltlombardi) [reply spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: They do not, but we can be sure to update once the constants move.

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb (hunk 1 above)
Comment (michaeltlombardi) [reply lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Addressed with a minor readme update:

File: README.md [hunk 6]
```diff
@@ -228,6 +228,7 @@ For all triggers:
   You should format dates as YYYY-MM-DD, although other date formats may work (under the hood, this uses Date.parse).
 * `minutes_interval` — The repeat interval in minutes.
 * `minutes_duration` — The duration in minutes, needs to be greater than the minutes_interval.
+* `random_minutes_interval` — The delay time that is randomly added to the start time of the trigger.
```
Comment (glennsarti) [inline README.md]: Is the default 0?

Comment (glennsarti) [reply README.md]: Can we use 0?

Comment (michaeltlombardi) [reply README.md]: Yes, the default is zero. I haven't tried to set it to zero (the default value is actually an empty string which it treats as a zero). Will test that now!

Comment (michaeltlombardi) [reply README.md]: Ensured the unit tests verify setting zero - it turns out that if you try setting it to zero it returns an empty string.

File: lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task.rb [hunk 7]
```diff
@@ -0,0 +1,260 @@
+# This class is used to manage V1 compatible tasks using the Task Scheduler V2 API
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task.rb]: This blurb is not right upon the 2nd commit and should get a minor update to reflect that.

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    it 'should be able to update a command' do
+      new_application_name = 'notepad.exe'
+      ps_cmd = '[string]((Get-ScheduledTask | ? { $_.TaskName -eq \'' + @task_name + '\' }).Actions[0].Execute)'
+      task = subject.activate(@task_name)
+
+      expect('cmd.exe').to be_same_as_powershell_command(ps_cmd)
+
+      subject.application_name = new_application_name
+      subject.save
+
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: whitespace

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    it 'should be able to update command parameters' do
+      new_parameters = '/nonsense /utter /nonsense'
+      ps_cmd = '[string]((Get-ScheduledTask | ? { $_.TaskName -eq \'' + @task_name + '\' }).Actions[0].Arguments)'
+      task = subject.activate(@task_name)
+
+      expect('/c exit 0').to be_same_as_powershell_command(ps_cmd)
+
+      subject.parameters = new_parameters
+      subject.save
+
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: remove whitespace

File: spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    it 'should be able to update the working directory' do
+      new_working_directory = 'C:\Somewhere'
+      ps_cmd = '[string]((Get-ScheduledTask | ? { $_.TaskName -eq \'' + @task_name + '\' }).Actions[0].WorkingDirectory)'
+      task = subject.activate(@task_name)
+
+      expect('').to be_same_as_powershell_command(ps_cmd)
+
+      subject.working_directory = new_working_directory
+      subject.save
+
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/scheduled_task/taskscheduler2_task_spec.rb]: remove whitespace

**Pull request merged**